examples/vae_simple.py

```diff
@@ -0,0 +1,322 @@
+            for mu, ls in zip(x_mu(b), x_ls(b))])/L
+        test_prediction = T.sum(x_mu(True), axis=0)/L
+    loss = -1 * (logpxz(False) + kl_div(False))
+    test_loss = -1 * (logpxz(True) + kl_div(True))
+
+    # functions for generating images given a code (used for visualization)
+    # we give the function a certain z and then set logsigma(z) to a very negative
+    # value to make the sampling nearly deterministic
+    z_var = T.dvector()
+    x_bin = nn.layers.get_output(l_x, {l_z_mu:z_var, l_z_ls:[-100,-100]},
+            deterministic=True)
+    x_cont = T.sum([nn.layers.get_output(l_x_mu, {l_z_mu:z_var, l_z_ls:[-100,-100]},
+            determistic=True) for l_x_mu in l_x_mu_list], axis=0) / L
+    gen_fn = theano.function([z_var], x_bin if binary else x_cont)
+        
+    # ADAM updates
+    params = nn.layers.get_all_params(l_x, trainable=True)
+    updates = nn.updates.adam(loss, params, learning_rate=1e-4)
+    train_fn = theano.function([input_var, target_var], loss, updates=updates)
+    val_fn = theano.function([input_var, target_var], test_loss)
+
+    print("Starting training...")
+    batch_size = 100
+    for epoch in range(num_epochs):
+        train_err = 0
+        train_batches = 0
+        start_time = time.time()
+        for batch in iterate_minibatches(X_train, X_train_tgt, batch_size, shuffle=True):
+            inputs, targets = batch
+            this_err = train_fn(inputs, targets)
+            train_err += this_err
+            train_batches += 1
+        val_err = 0
+        val_batches = 0
+        for batch in iterate_minibatches(X_val, X_val_tgt, batch_size, shuffle=False):
+            inputs, targets = batch
+            err = val_fn(inputs, targets)
+            val_err += err
+            val_batches += 1
+        print("Epoch {} of {} took {:.3f}s".format(
+            epoch + 1, num_epochs, time.time() - start_time))
+        print("  training loss:\t\t{:.6f}".format(train_err / train_batches))
+        print("  validation loss:\t\t{:.6f}".format(val_err / val_batches))
+
+    test_err = 0
+    test_batches = 0
+    for batch in iterate_minibatches(X_test, X_test_tgt, batch_size, shuffle=False):
+        inputs, targets = batch
+        err = val_fn(inputs, targets)
+        test_err += err
+        test_batches += 1
+    test_err /= test_batches
+    print("Final results:")
+    print("  test loss:\t\t\t{:.6f}".format(test_err))
+    # save some example pictures so we can see what it's done 
+    example_batch_size = 20
+    X_comp = X_test[:example_batch_size]
+    pred_fn = theano.function([input_var], test_prediction)
+    X_pred = pred_fn(X_comp).reshape(-1, 1, width, height)
+    for i in range(20):
+        get_image_pair(X_comp, X_pred, idx=i, channels=1).save('output_{}.jpg'.format(i))
+    # save the parameters so they can be loaded for next time
+    print("Saving")
+    fn = 'p_{:.6f}.params'.format(test_err)
+    pickle.dump(nn.layers.get_all_param_values(l_x), open(fn, 'w'))
```

Comment:
I'd add `protocol=-1` in the end to get a much smaller file in much shorter time.